File: ugoite-cli/tests/test_space.py
"""Tests for space management.

REQ-STO-001: Storage abstraction using fsspec.
REQ-STO-002: Space directory structure.
REQ-STO-003: File permissions.
REQ-STO-004: Space management via global.json.
REQ-STO-005: Prevent duplicate space creation.
"""

import json

import fsspec
import pytest

from ugoite.space import (
    SampleSpaceOptions,
    SpaceExistsError,
    create_sample_space,
    create_space,
)
from ugoite.utils import fs_join


def test_create_space_scaffolding(
    fs_impl: tuple[fsspec.AbstractFileSystem, str],
) -> None:
    """Verifies that create_space generates the required file structure.

    It also verifies metadata files as per Spec 03 §2.
    """
    fs, root = fs_impl
    ws_id = "test-space"

    # Call the function to be implemented
    create_space(root, ws_id, fs=fs)

    # Verify directory structure
    ws_path = fs_join(root, "spaces", ws_id)
    assert fs.exists(ws_path)
    assert fs.exists(fs_join(ws_path, "meta.json"))
    assert fs.exists(fs_join(ws_path, "settings.json"))
    assert fs.exists(fs_join(ws_path, "forms"))
    assert fs.exists(fs_join(ws_path, "assets"))

    # Verify global.json exists at root
    global_json_path = fs_join(root, "global.json")
    assert fs.exists(global_json_path)

    # Verify global.json content includes the new space
    with fs.open(global_json_path, "r") as f:
        global_data = json.load(f)
        assert "spaces" in global_data
        assert ws_id in global_data["spaces"]

    # Verify meta.json content
    with fs.open(fs_join(ws_path, "meta.json"), "r") as f:
        meta = json.load(f)
        assert meta["id"] == ws_id
        assert meta["name"] == ws_id
        assert "created_at" in meta
        assert "storage" in meta

    # Permission enforcement is handled by the core storage layer.


def test_create_space_idempotency(
    fs_impl: tuple[fsspec.AbstractFileSystem, str],
) -> None:
    """Verifies that creating an existing space raises SpaceExistsError."""
    fs, root = fs_impl
    ws_id = "test-space"
    create_space(root, ws_id, fs=fs)

    # Should raise error if exists
    with pytest.raises(SpaceExistsError):
        create_space(root, ws_id, fs=fs)


def test_create_space_s3_unimplemented() -> None:
    """Verifies that providing a non-local URI (e.g. s3://) raises NotImplementedError.

    This is for Milestone 0.
    """
    root = "s3://my-bucket/ugoite_root"
    ws_id = "test-space"

    # We expect this to fail because we only support local fs in M0
    # or we haven't configured S3 credentials/mocking.
    # But the requirement is to "raise unimplemented error".
    with pytest.raises(NotImplementedError):
        create_space(root, ws_id)


def test_create_sample_space_req_api_009(
    fs_impl: tuple[fsspec.AbstractFileSystem, str],
) -> None:
    """REQ-API-009: CLI can generate a sample-data space."""
    fs, root = fs_impl
    ws_id = "sample-space"

    options = SampleSpaceOptions(
        scenario="renewable-ops",
        entry_count=120,
        seed=123,
    )
    summary = create_sample_space(root, ws_id, options=options, fs=fs)

    assert summary["space_id"] == ws_id
    assert summary["entry_count"] == 120
    assert 3 <= summary["form_count"] <= 6

File: ugoite-cli/tests/test_space.rs
//! Integration tests for space management commands.
//! REQ-STO-001, REQ-STO-002, REQ-STO-003, REQ-STO-004, REQ-STO-005, REQ-API-009

use std::process::Command;

fn ugoite_bin() -> String {
    let mut path = std::env::current_exe().unwrap();
    path.pop();
    if path.ends_with("deps") {
        path.pop();
    }
    path.push("ugoite");
    path.to_string_lossy().to_string()
}

#[cfg(unix)]
fn mode(path: &std::path::Path) -> u32 {
    use std::os::unix::fs::PermissionsExt;

    std::fs::metadata(path).unwrap().permissions().mode() & 0o777
}

/// REQ-STO-001, REQ-STO-002: Create space scaffolding at local path.
#[test]
fn test_create_space_scaffolding() {
    let dir = tempfile::tempdir().unwrap();
    let root = dir.path().to_string_lossy().to_string();
    let config_path = dir.path().join("cli-config.json");

    let output = Command::new(ugoite_bin())
        .args(["create-space", "--root", &root, "my-space"])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("failed to execute");

    assert!(
        output.status.success(),
        "stderr: {}",
        String::from_utf8_lossy(&output.stderr)
    );

    // Verify the space directory was created
    let space_dir = dir.path().join("spaces").join("my-space");
    assert!(space_dir.exists(), "Space directory should be created");
}

#[cfg(unix)]
/// REQ-STO-003: Create space applies owner-only local permissions.
#[test]
fn test_create_space_req_sto_003_permissions() {
    let dir = tempfile::tempdir().unwrap();
    let root = dir.path().to_string_lossy().to_string();
    let config_path = dir.path().join("cli-config.json");

    let output = Command::new(ugoite_bin())
        .args(["create-space", "--root", &root, "private-space"])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("failed to execute");

    assert!(
        output.status.success(),
        "stderr: {}",
        String::from_utf8_lossy(&output.stderr)
    );

    let spaces_root = dir.path().join("spaces");
    let space_dir = spaces_root.join("private-space");
    assert_eq!(mode(&spaces_root), 0o700);
    assert_eq!(mode(&space_dir), 0o700);
    for dir_name in ["forms", "assets", "materialized_views", "sql_sessions"] {
        assert_eq!(mode(&space_dir.join(dir_name)), 0o700);
    }
    for file_name in ["meta.json", "settings.json"] {
        assert_eq!(mode(&space_dir.join(file_name)), 0o600);
    }
}

/// REQ-STO-001: S3 backend is not yet implemented (returns unimplemented error).
#[test]
fn test_create_space_s3_unimplemented() {
    let dir = tempfile::tempdir().unwrap();
    let config_path = dir.path().join("cli-config.json");

    // Attempting to use s3:// path should fail gracefully
    let output = Command::new(ugoite_bin())
        .args(["create-space", "--root", "s3://my-bucket", "my-space"])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("failed to execute");

    // S3 is not supported in core mode; expect failure
    assert!(
        !output.status.success() || {
            let stderr = String::from_utf8_lossy(&output.stderr);
            stderr.contains("not") || stderr.contains("unsupported") || stderr.contains("error")
        }
    );
}

/// REQ-STO-005: Prevent duplicate space creation - returns error for existing space.
#[test]
fn test_create_space_idempotency() {
    let dir = tempfile::tempdir().unwrap();
    let root = dir.path().to_string_lossy().to_string();
    let config_path = dir.path().join("cli-config.json");

    // Create space first time
    Command::new(ugoite_bin())
        .args(["create-space", "--root", &root, "idempotent-space"])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("failed to execute");

    // Second creation should fail with "already exists" error
    let output2 = Command::new(ugoite_bin())
        .args(["create-space", "--root", &root, "idempotent-space"])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("failed to execute");

    assert!(
        !output2.status.success(),
        "Second create-space should fail for duplicate space"
    );
    let stderr = String::from_utf8_lossy(&output2.stderr);
    assert!(
        stderr.contains("already exists"),
        "Expected 'already exists' error, got: {stderr}"
    );
}

/// REQ-API-009: Sample space can be created with sample data.
#[test]
fn test_create_sample_space_req_api_009() {
    let dir = tempfile::tempdir().unwrap();
    let root = dir.path().to_string_lossy().to_string();
    let config_path = dir.path().join("cli-config.json");

    let output = Command::new(ugoite_bin())
        .args(["create-space", "--root", &root, "sample-space"])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("failed to execute");

    assert!(
        output.status.success(),
        "stderr: {}",
        String::from_utf8_lossy(&output.stderr)
    );

    // Verify space was created
    let space_dir = dir.path().join("spaces").join("sample-space");
    assert!(space_dir.exists());
}

/// REQ-API-009: Direct sample-data CLI should show progress and create the target space.
#[test]
fn test_sample_data_progress_req_api_009() {
    let dir = tempfile::tempdir().unwrap();
    let root = dir.path().to_string_lossy().to_string();
    let config_path = dir.path().join("cli-config.json");

    let output = Command::new(ugoite_bin())
        .args([
            "space",
            "sample-data",
            &root,
            "sample-progress",
            "--scenario",
            "lab-qa",
            "--entry-count",
            "10",
            "--seed",
            "7",
        ])
        .env("UGOITE_CLI_CONFIG_PATH", &config_path)
        .output()
        .expect("failed to execute");

    assert!(
        output.status.success(),
        "stderr: {}",
        String::from_utf8_lossy(&output.stderr)
    );

    let stdout = String::from_utf8_lossy(&output.stdout);
    assert!(
        stdout.contains("\"created\": true"),
        "Expected created JSON output, got: {stdout}"
    );

    let stderr = String::from_utf8_lossy(&output.stderr);
    assert!(
        stderr.contains("Seed progress ["),
        "Expected progress output, got: {stderr}"
    );
    assert!(
        stderr.contains("(10/10) Completed"),
        "Expected completed progress output, got: {stderr}"
    );

    let space_dir = dir.path().join("spaces").join("sample-progress");
    assert!(
        space_dir.exists(),
        "Sample data command should create the space"
    );
}
